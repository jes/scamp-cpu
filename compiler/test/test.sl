# Test program for SLANG

include "stdio.sl";
include "stdlib.sl";
include "string.sl";
include "list.sl";

# recursive fibonacci
var rfib = func(n) {
    if (n <= 2)
        return 1
    else
        return rfib(n-1)+rfib(n-2);
};

# reverse the given string in-place
var strrev = func(s) {
    var ss = s;
    var p = s;

    while (*(p+1))
        p++;

    var ch;

    while (s < p) {
        ch = *p;
        *p = *s;
        *s = ch;
        --p;
        ++s;
    };

    return ss;
};

var optest = func() {
    var a = 0xaaaa;
    var b = 0x55aa;

    puts("Operators test:\n");

    printf("a=%d\n", [a]);
    printf("b=%d\n", [b]);
    printf("~a=%d\n", [~a]);
    printf("-b=%d\n", [-b]);
    printf("+a=%d\n", [+a]);
    printf("a+b=%d\n", [a+b]);
    printf("a-b=%d\n", [a-b]);
    printf("a&b=%d\n", [a&b]);
    printf("a|b=%d\n", [a|b]);
    printf("a^b=%d\n", [a^b]);
    printf("a&&b=%d\n", [a&&b]);
    printf("a||b=%d\n", [a||b]);
    printf("a&&0=%d\n", [a&&0]);
    printf("a||0=%d\n", [a||0]);
    printf("(-1 > 0)=%d\n", [-1>0]);
    printf("(-1 >= 0)=%d\n", [-1>=0]);
    printf("(-1 < 0)=%d\n", [-1<0]);
    printf("(-1 <= 0)=%d\n", [-1<=0]);
    printf("(32767>0)=%d\n", [32767>0]);
    printf("(32767>=0)=%d\n", [32767>=0]);
    printf("(32767<0)=%d\n", [32767<0]);
    printf("(32767<=0)=%d\n", [32767<=0]);
    printf("(32767>-2)=%d\n", [32767>-2]);
    printf("(32767>=-2)=%d\n", [32767>=-2]);
    printf("(32767<-2)=%d\n", [32767<-2]);
    printf("(32767<=-2)=%d\n", [32767<=-2]);

    var x = 5;
    var y = 10;
    printf("x=%d\n", [x]);
    printf("y=%d\n", [y]);
    printf("x>y-6=%d\n", [x>y-6]);

    puts("\n");
    puts("unsigned magnitude comparison:\n");
    a = 0;
    while (a != 0xff00) {
        b = a;
        while (b != 0xff00) {
            printf("%d < %d = %d\n", [a, b, a < b]);
            printf("%d > %d = %d\n", [a, b, a > b]);
            printf("%d <= %d = %d\n", [a, b, a <= b]);
            printf("%d >= %d = %d\n", [a, b, a >= b]);

            printf("%d lt %d = %d\n", [a, b, a lt b]);
            printf("%d gt %d = %d\n", [a, b, a gt b]);
            printf("%d le %d = %d\n", [a, b, a le b]);
            printf("%d ge %d = %d\n", [a, b, a ge b]);

            b = b + 16320;
        };
        a = a + 16320;
    };

    puts("\n");
};

var fibtest = func() {
    var n = 0;
    var f1 = 1;
    var f2 = 0;
    var f3 = 0;

    puts("Fibonacci numbers:\n");

    while (n++ <= 22) {
        f3 = f1 + f2;
        f1 = f2;
        f2 = f3;

        printf("fib(%d) = %d\n", [n,f3]);
        if (n <= 18) printf("rfib(%d) = %d\n", [n,rfib(n)]);
    };

    puts("\n");
};

var strtest = func() {
    var s1 = "Hello, world!";

    printf("string: %s\n", [s1]);
    printf("reversed: %s\n", [strrev(s1)]);
    printf("re-reversed: %s\n", [strrev(s1)]);

    printf("cmp self: %d\n", [strcmp(s1,s1)]);
    printf("cmp hello world: %d\n", [strcmp(s1,"Hello, world!")]);
    printf("cmp foo: %d\n", [strcmp(s1,"foo")]);

    printf("length: %d\n", [strlen(s1)]);

    puts("\n");
};

var callputs = func(f) {
    puts(f());
};

var globule = 1;

var functest = func() {
    puts("Nested function calls/declarations:\n");

    callputs(func() { return "1. Hello, world\n" });

    var f = func(x) {
        return x() + x();
    };

    printf("2. 5 = %d\n", [f(func() {
        globule++;
        return globule;
    })]);

    # this function produces no output, but will leave the sp incorrect if
    # vars inside the function body interact poorly with early returns
    var inlinevars = func(x) {
        var y = x;
        if (y > 100)
            return 1;
        var z = x+2;
        if (z > 100)
            return 1;
        return 0;
    };

    inlinevars(97);
    inlinevars(99);
    inlinevars(101);

    puts("\n");
};

var ptrtest = func() {
    var x = 5;
    var p = &x;
    var y = 10;

    puts("Pointers to locals:\n");

    printf("1. x = %d\n", [x]);
    printf("1. y = %d\n", [y]);
    printf("1. *p = %d\n", [*p]);
    *p = 7;
    printf("2. x = %d\n", [x]);
    printf("2. y = %d\n", [y]);
    printf("2. *p = %d\n", [*p]);
    p = &y;
    printf("3. x = %d\n", [x]);
    printf("3. y = %d\n", [y]);
    printf("3. *p = %d\n", [*p]);
    *p = 11;
    printf("4. x = %d\n", [x]);
    printf("4. y = %d\n", [y]);
    printf("4. *p = %d\n", [*p]);

    puts("\n");
};

var looptest = func() {
    var i = 0;

    puts("loop break/continue:\n");

    while (i++ < 100) {
        puts(itoa(i));
        puts(": ");
        if (i >= 20) {
            puts("i >= 20\n");
            break;
        };
        if (i >= 10) {
            puts("i >= 10\n");
            continue;
        };
        puts("i < 10\n");
    };

    puts("\n");
};

var fizzbuzz = func() {
    var count3 = 1;
    var count5 = 1;
    var i = 0;

    puts("fizzbuzz:\n");

    while (++i < 100) {
        if (count3==0 && count5==0)
            puts("fizzbuzz")
        else if (count3==0)
            puts("fizz")
        else if (count5==0)
            puts("buzz")
        else
            puts(itoa(i));
        puts(" ");

        count3 = count3+1;
        if (count3 == 3)
            count3 = 0;
        count5 = count5+1;
        if (count5 == 5)
            count5 = 0;
    };

    puts("\n");
    puts("\n");
};

var primestest = func() {
    var maxprime = 1000;
    var prime = malloc(maxprime);
    memset(prime, 1, maxprime);

    # output the difference between "prime" and the next malloc, to highlight any changes
    var x = malloc(1);
    printf("x-prime=%d\n\n", [x-prime]);

    puts("primes:\n");

    *(prime+0) = 0;
    *(prime+1) = 0;
    *(prime+2) = 1;

    var i = 1;
    var isqr;
    var j;
    while (i < maxprime) {
        if (*(prime+i)) {
            j = i+i;
            while (j < maxprime) {
                *(prime+j) = 0;
                j = j + i;
            };
        };
        i = i + 2;
    };

    puts("2");
    i = 3;
    while (i < maxprime) {
        if (*(prime+i)) printf(" %d", [i]);
        i = i + 2;
    };

    puts("\n");
    puts("\n");

    free(prime);
};

var l2;
var listtest = func() {
    var l = lstnew();

    puts("list:\n");

    lstpush(l, 3);
    lstpush(l, 4);
    lstunshift(l, 2);
    lstunshift(l, 1);

    lstwalk(l, func(v) { puts(itoa(v)); puts(" "); });
    puts("\nlength = "); puts(itoa(lstlen(l)));
    puts("\npop "); puts(itoa(lstpop(l)));
    puts("\nlength = "); puts(itoa(lstlen(l))); puts("\n");

    # reverse l into l2 (l2 has to be a global because the callback can't access
    # the local stack frame)
    l2 = lstnew();
    lstwalk(l, func(v) {
        lstunshift(l2, v);
    });

    lstwalk(l2, func(v) { puts(itoa(v)); puts(" "); });
    puts("\nlength = "); puts(itoa(lstlen(l2)));
    puts("\nshift "); puts(itoa(lstshift(l2)));
    puts("\nlength = "); puts(itoa(lstlen(l2))); puts("\n");

    puts("\n");

    lstfree(l);
    lstfree(l2);
};

var arraytest = func() {
    puts("array:\n");
    var arr;
    var p;
    var i = 0;
    while (i++ < 10) {
        arr = ["Hello, ", "world! ", "For time no. ", itoa(i), "\n"];
        p = arr;
        while (*p) {
            puts(*(p++));
        };

        puts("arr[3]="); puts(arr[3]); puts("\n");
        puts("strcmp(arr[0],\"Hello, \")="); puts(itoa(strcmp(arr[0], "Hello, "))); puts("\n");
        puts("strcmp(arr[0],\"foo\")="); puts(itoa(strcmp(arr[0], "foo"))); puts("\n");
    };
    puts("\n");
};

var sp = 0xffff;
var initial_sp = *sp;
optest();
fibtest();
strtest();
functest();
ptrtest();
looptest();
fizzbuzz();
primestest();
listtest();
arraytest();

var new_sp = *sp;
puts("sp change="); puts(itoa(new_sp-initial_sp)); puts("\n");
