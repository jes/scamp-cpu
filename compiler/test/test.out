Operators test:
a=43690
b=21930
~a=21845
-b=43606
+a=43690
a+b=84
a-b=21760
a&b=170
a|b=65450
a^b=65280
a&&b=1
a||b=1
a&&0=0
a||0=1
(-1 > 0)=0
(-1 >= 0)=0
(-1 < 0)=1
(-1 <= 0)=1
(32767>0)=1
(32767>=0)=1
(32767<0)=0
(32767<=0)=0
(32767>-2)=1
(32767>=-2)=1
(32767<-2)=0
(32767<=-2)=0
x=5
y=10
x>y-6=1

unsigned magnitude comparison:
0 < 0 = 0
0 > 0 = 0
0 <= 0 = 1
0 >= 0 = 1
0 lt 0 = 0
0 gt 0 = 0
0 le 0 = 1
0 ge 0 = 1
0 < 16320 = 1
0 > 16320 = 0
0 <= 16320 = 1
0 >= 16320 = 0
0 lt 16320 = 1
0 gt 16320 = 0
0 le 16320 = 1
0 ge 16320 = 0
0 < 32640 = 1
0 > 32640 = 0
0 <= 32640 = 1
0 >= 32640 = 0
0 lt 32640 = 1
0 gt 32640 = 0
0 le 32640 = 1
0 ge 32640 = 0
0 < 48960 = 0
0 > 48960 = 1
0 <= 48960 = 0
0 >= 48960 = 1
0 lt 48960 = 1
0 gt 48960 = 0
0 le 48960 = 1
0 ge 48960 = 0
16320 < 16320 = 0
16320 > 16320 = 0
16320 <= 16320 = 1
16320 >= 16320 = 1
16320 lt 16320 = 0
16320 gt 16320 = 0
16320 le 16320 = 1
16320 ge 16320 = 1
16320 < 32640 = 1
16320 > 32640 = 0
16320 <= 32640 = 1
16320 >= 32640 = 0
16320 lt 32640 = 1
16320 gt 32640 = 0
16320 le 32640 = 1
16320 ge 32640 = 0
16320 < 48960 = 0
16320 > 48960 = 1
16320 <= 48960 = 0
16320 >= 48960 = 1
16320 lt 48960 = 1
16320 gt 48960 = 0
16320 le 48960 = 1
16320 ge 48960 = 0
32640 < 32640 = 0
32640 > 32640 = 0
32640 <= 32640 = 1
32640 >= 32640 = 1
32640 lt 32640 = 0
32640 gt 32640 = 0
32640 le 32640 = 1
32640 ge 32640 = 1
32640 < 48960 = 0
32640 > 48960 = 1
32640 <= 48960 = 0
32640 >= 48960 = 1
32640 lt 48960 = 1
32640 gt 48960 = 0
32640 le 48960 = 1
32640 ge 48960 = 0
48960 < 48960 = 0
48960 > 48960 = 0
48960 <= 48960 = 1
48960 >= 48960 = 1
48960 lt 48960 = 0
48960 gt 48960 = 0
48960 le 48960 = 1
48960 ge 48960 = 1

Fibonacci numbers:
fib(1) = 1
rfib(1) = 1
fib(2) = 1
rfib(2) = 1
fib(3) = 2
rfib(3) = 2
fib(4) = 3
rfib(4) = 3
fib(5) = 5
rfib(5) = 5
fib(6) = 8
rfib(6) = 8
fib(7) = 13
rfib(7) = 13
fib(8) = 21
rfib(8) = 21
fib(9) = 34
rfib(9) = 34
fib(10) = 55
rfib(10) = 55
fib(11) = 89
rfib(11) = 89
fib(12) = 144
rfib(12) = 144
fib(13) = 233
rfib(13) = 233
fib(14) = 377
rfib(14) = 377
fib(15) = 610
rfib(15) = 610
fib(16) = 987
rfib(16) = 987
fib(17) = 1597
rfib(17) = 1597
fib(18) = 2584
rfib(18) = 2584
fib(19) = 4181
fib(20) = 6765
fib(21) = 10946
fib(22) = 17711
fib(23) = 28657

string: Hello, world!
reversed: !dlrow ,olleH
re-reversed: Hello, world!
cmp self: 0
cmp hello world: 0
cmp foo: 65506
length: 13

Nested function calls/declarations:
1. Hello, world
2. 5 = 5

Pointers to locals:
1. x = 5
1. y = 10
1. *p = 5
2. x = 7
2. y = 10
2. *p = 7
3. x = 7
3. y = 10
3. *p = 10
4. x = 7
4. y = 11
4. *p = 11

loop break/continue:
1: i < 10
2: i < 10
3: i < 10
4: i < 10
5: i < 10
6: i < 10
7: i < 10
8: i < 10
9: i < 10
10: i >= 10
11: i >= 10
12: i >= 10
13: i >= 10
14: i >= 10
15: i >= 10
16: i >= 10
17: i >= 10
18: i >= 10
19: i >= 10
20: i >= 20

fizzbuzz:
1 2 fizz 4 buzz fizz 7 8 fizz buzz 11 fizz 13 14 fizzbuzz 16 17 fizz 19 buzz fizz 22 23 fizz buzz 26 fizz 28 29 fizzbuzz 31 32 fizz 34 buzz fizz 37 38 fizz buzz 41 fizz 43 44 fizzbuzz 46 47 fizz 49 buzz fizz 52 53 fizz buzz 56 fizz 58 59 fizzbuzz 61 62 fizz 64 buzz fizz 67 68 fizz buzz 71 fizz 73 74 fizzbuzz 76 77 fizz 79 buzz fizz 82 83 fizz buzz 86 fizz 88 89 fizzbuzz 91 92 fizz 94 buzz fizz 97 98 fizz 

x-prime=65532

primes:
2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97 101 103 107 109 113 127 131 137 139 149 151 157 163 167 173 179 181 191 193 197 199 211 223 227 229 233 239 241 251 257 263 269 271 277 281 283 293 307 311 313 317 331 337 347 349 353 359 367 373 379 383 389 397 401 409 419 421 431 433 439 443 449 457 461 463 467 479 487 491 499 503 509 521 523 541 547 557 563 569 571 577 587 593 599 601 607 613 617 619 631 641 643 647 653 659 661 673 677 683 691 701 709 719 727 733 739 743 751 757 761 769 773 787 797 809 811 821 823 827 829 839 853 857 859 863 877 881 883 887 907 911 919 929 937 941 947 953 967 971 977 983 991 997

list:
1 2 3 4 
length = 4
pop 4
length = 3
3 2 1 
length = 3
shift 3
length = 2

array:
Hello, world! For time no. 1
arr[3]=1
strcmp(arr[0],"Hello, ")=0
strcmp(arr[0],"foo")=65506
Hello, world! For time no. 2
arr[3]=2
strcmp(arr[0],"Hello, ")=0
strcmp(arr[0],"foo")=65506
Hello, world! For time no. 3
arr[3]=3
strcmp(arr[0],"Hello, ")=0
strcmp(arr[0],"foo")=65506
Hello, world! For time no. 4
arr[3]=4
strcmp(arr[0],"Hello, ")=0
strcmp(arr[0],"foo")=65506
Hello, world! For time no. 5
arr[3]=5
strcmp(arr[0],"Hello, ")=0
strcmp(arr[0],"foo")=65506
Hello, world! For time no. 6
arr[3]=6
strcmp(arr[0],"Hello, ")=0
strcmp(arr[0],"foo")=65506
Hello, world! For time no. 7
arr[3]=7
strcmp(arr[0],"Hello, ")=0
strcmp(arr[0],"foo")=65506
Hello, world! For time no. 8
arr[3]=8
strcmp(arr[0],"Hello, ")=0
strcmp(arr[0],"foo")=65506
Hello, world! For time no. 9
arr[3]=9
strcmp(arr[0],"Hello, ")=0
strcmp(arr[0],"foo")=65506
Hello, world! For time no. 10
arr[3]=10
strcmp(arr[0],"Hello, ")=0
strcmp(arr[0],"foo")=65506

sp change=0
